### ISSUE TITLE
Singularity image built from poldracklab/fmriprep:latest requires additional ldconfig

### ISSUE THREAD

**Issue by justbennet** [poldracklab/fmriprep#749]
In Issue 648 for mriqc, https://github.com/poldracklab/mriqc/issues/648#issuecomment-335242120, I reported that when I converted to Singularity from Docker, there were problems because libGL.so.1 is missing.  This seems also to be an issue with fmriprep.

I used the following to create the Singularity container, where Singularity was compiled from the development branch, pre-3.4 release.

fmriprep.def
```
Bootstrap: docker
From: poldracklab/fmriprep:latest

%post
chmod -R a+rX /usr/local/miniconda
chmod -R a+rX /niworkflows_data
```
I am running tests against the `ds008` dataset from OpenfMRI.org.  When it got to
```
Node: fmriprep_wf.single_subject_15_wf.anat_preproc_wf.surface_recon_wf.fs_transform
```
it crashed when the following error was returned to `niworkflows/nipype`.
```
RuntimeError: Command:
tkregister2 --fslregout freesurfer2subT1.mat --mov /out/freesurfer/sub-15/mri/T1.mgz --noedit --reg register.dat --regheader --targ /home/bennet/Projects/mpeckins/MRIPREP/work/fmriprep_wf/single_subject_15_wf/anat_preproc_wf/t1_merge/sub-15_T1w_template.nii.gz
Standard output:

Standard error:
/opt/freesurfer/tktools/tkregister2.bin: error while loading shared libraries: libGL.so.1: cannot open shared object file: No such file or directory

Return code: 127
Interface Tkregister2 failed to run. 
```
Using `singularity shell` in both containers, and running `dkpg -l`, searching for either GL or mesa fails in both containers

```
Singularity mriqc.simg:/> dpkg -l | grep GL
Singularity mriqc.simg:/> dpkg -l | grep mesa

Singularity fmriprep.simg:/> dpkg -l | grep GL
Singularity fmriprep.simg:/> dpkg -l | grep mesa
```
In both containers, the files appear to be there, though.
```
Singularity mriqc.simg:/> find /usr/lib/x86_64-linux-gnu -name libGL\*
/usr/lib/x86_64-linux-gnu/libGL.so.1
/usr/lib/x86_64-linux-gnu/libGLU.so.1
/usr/lib/x86_64-linux-gnu/libGLU.so.1.3.1
/usr/lib/x86_64-linux-gnu/libGLw.so.1
/usr/lib/x86_64-linux-gnu/libGLw.so.1.0.0
/usr/lib/x86_64-linux-gnu/mesa/libGL.so.1
/usr/lib/x86_64-linux-gnu/mesa/libGL.so.1.2.0

Singularity fmriprep.simg:/> find /usr/lib/x86_64-linux-gnu -name libGL\*
/usr/lib/x86_64-linux-gnu/libGLU.so.1
/usr/lib/x86_64-linux-gnu/libGLU.so.1.3.1
/usr/lib/x86_64-linux-gnu/libGLw.so.1
/usr/lib/x86_64-linux-gnu/libGLw.so.1.0.0
/usr/lib/x86_64-linux-gnu/mesa/libGL.so.1
/usr/lib/x86_64-linux-gnu/mesa/libGL.so.1.2.0
```
I don't see anything that explicitly says to install any libGL libraries, so I presume that these are getting installed somehow as a dependency for something else, possibly (probably) FSL.  Whatever it is, it seems that the mesa library is not added to the library search path?
```
Singularity fmriprep.simg:~/Projects/mpeckins/FMRIPREP> ldconfig -p | grep libGL
Singularity fmriprep.simg:~/Projects/mpeckins/FMRIPREP> ldconfig -p | grep mesa
Singularity fmriprep.simg:~/Projects/mpeckins/FMRIPREP> echo $LD_LIBRARY_PATH
/usr/lib/fsl/5.0::/.singularity.d/libs
```

After trying several other manual fixes, I trying it after adding `ldconfig` in the `%post` section of the Singularity image definition file.  It will take a while to run, but that does cause the library to appear in the library path.
```
$ singularity shell fmriprep-fixed.simg
Singularity: Invoking an interactive shell within container...

Singularity fmriprep-fixed.simg:~/Projects/mpeckins/FMRIPREP> ldconfig -p | grep mesa
	libGL.so.1 (libc6,x86-64) => /usr/lib/x86_64-linux-gnu/mesa/libGL.so.1
	libEGL.so.1 (libc6,x86-64) => /usr/lib/x86_64-linux-gnu/mesa-egl/libEGL.so.1
```

This may not be germane to the main issues, but it
```
apt-get clean && rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*
```
doing such a good job that it removes from the dpkg database the results of the following command, which is where I would surmise the GL libraries come in as a dependency?
```
    apt-get install -y --no-install-recommends \
                    fsl-core=5.0.9-4~nd16.04+1 \
                    fsl-mni152-templates=5.0.7-2 \
                    afni=16.2.07~dfsg.1-5~nd16.04+1
```
This
```
Singularity fmriprep.simg:/> dpkg -l | grep fsl
```
returns empty.

**Comment by effigies** [comment]
Singularity's a little bit of an odd beast, since it may be showing you files on your own filesystem, which may be why `ldconfig` is working from inside the container.

Anyway, can you give your commands for creating the singularity image? I've had success with `docker2singularity`, but I haven't been keeping up with the latest in the singularity world, and can't replicate without specific commands.

**Comment by justbennet** [comment]
Chris,

I think all containers are bit odd.  For the ldconfig, it failed to show the libraries before I added the `ldconfig` command.  The host isn't changing in between, so I'm fairly confident that it isn't finding those.  I don't think Singularity exposes the library directories to the container, but I will try to confirm that.  Thanks for bringing that up.

**Comment by effigies** [comment]
Hi @justbennet, have you made any progress on figuring out this issue? Anything we can help with?

**Comment by justbennet** [comment]
It appears that adding an ldconfig to the %post section of the Singularity recipe solves the problem.  My original Singularity recipe is at the top of the issue; here is the one that produces a fully-functional Singularity container.
```
Bootstrap: docker
From: poldracklab/fmriprep:latest

%setup

%post
chmod -R a+rX /usr/local/miniconda
chmod -R a+rX /niworkflows_data
ldconfig
```
I will retest both of these with the just-released 3.4 version of Singularity in the next few days to confirm.  It's possible that a later commit to Singularity solves the issue.

Please also note that additional permission changes seem to be needed.  I submitted a PR for the miniconda change to MRIQC, which was accepted.  I can create a PR for those here, too, if you think those a good addition.  I did get permission denied in both places initially.

**Comment by effigies** [comment]
Ah, nice. I didn't see those above. I'll make a PR to update those two directories.

Thanks for looking into this. Incidentally, have you tried running those three commands at the end of the Dockerfile, and having nothing in your `%post`? If that works, we can definitely add `ldconfig` to our Dockerfile. But if it's a Singularity issue, then it might make more sense to host a stub recipe like the one you have here.

**Comment by justbennet** [comment]
Chris,

I don't have Docker.  I guess I'm the test case for the Docker-less user.  We're gradually hoping that using Singularity instead of Docker will make it easy and straightforward for us to create containers for our cluster.

There already is an `ldconfig` at line 168 in the Dockerfile that I'm surprised isn't doing the job.  I saw some traffic on the Singularity list about the effects of ordering of Docker layers, and I wonder if that might be it, but I don't know.  I also saw that some commands (e.g., conda) in the Docker file are followed by `sync` commands, perhaps and additional `sync` would be helpful?

If you have a way for me to test some change you make in the Docker image, let me know.

Thanks for your attention to this.  I hope I'm not too much a special case.

**Comment by effigies** [comment]
You're right. Not sure if `sync` would help. (@oesteban any thoughts?)

And I'm not sure what the beset way to allow you to test these options would be. We could probably make a one-time branch that pushes a test image to Docker Hub that you could try out.

**Comment by justbennet** [comment]
Offer's open.  Sorry I can't be more help on the Docker end of things.  Let me know if there is something I should try.  It might take me a day or two or three to fully test; got a pesky day job.  ;-)

**Comment by effigies** [comment]
Hi @justbennet, have you been keeping up with the latest releases? I just want to check in whether this is still an issue. I believe we covered the `chmod` steps, but I'm not sure that we resolved the `ldconfig` issue.

**Comment by yonestar** [comment]
Hi.  I followed the Singularity installation instruction provided on the readthedocs (i.e., using docker2singularity), and encountered a similar issue.  I no longer have the exact error message (sorry), but something along the lines of invalid ELF header file for a gnu library.  My guess is that it was a permissions issues.  I also tried running fMRIprep directly from Singularity, pulling from Docker hub, and received the same error.  What worked for me was creating a simple Singularity recipe file (posted [here](https://github.com/canlab/cantainer/blob/master/Singularity)) and building the Singularity image from that.  Note that I did not need an `ldconfig` in my recipe file.

So, the posted Singularity installation instructions did not work for me.  I don't know whether this is specific to my system, whether its an error in docker2singularity, or what... Sorry I can't be more informative; just sharing what I found.
